Public/Src/Cache/ContentStore/Library/FileSystem/PassThroughFileSystem.cs

```diff
@@ -445,89 +446,86 @@ private async Task CopyFileInsideSemaphoreAsync(AbsolutePath sourcePath, Absolut
             }
         }
 
-        private Stream OpenInternal(AbsolutePath path, FileAccess accessMode, FileMode mode, FileShare share, FileOptions options, int bufferSize)
+        private Stream TryOpenFile(AbsolutePath path, FileAccess accessMode, FileMode mode, FileShare share, FileOptions options, int bufferSize)
         {
-                if (OperatingSystemHelper.IsUnixOS)
-                {
-                    if (DirectoryExists(path))
-                    {
-                        throw new UnauthorizedAccessException($"Cannot open directory {path} as a file.");
-                    }
+            if (OperatingSystemHelper.IsUnixOS)
+            {
+                return TryOpenFileUnix(path, accessMode, mode, share, options, bufferSize);
+            }
 
-                    if (mode == FileMode.Open && !FileExists(path))
-                    {
-                        return null;
-                    }
+            return TryOpenFileWin(path, accessMode, mode, share, options, bufferSize);
+        }
 
-                    return new FileStream(path.Path, mode, accessMode, share);
-                }
+        private Stream TryOpenFileUnix(AbsolutePath path, FileAccess accessMode, FileMode mode, FileShare share, FileOptions options, int bufferSize)
+        {
+            if (DirectoryExists(path))
+            {
+                throw new UnauthorizedAccessException($"Cannot open directory {path} as a file.");
+            }
 
-                int access = GetDwAccess(accessMode);
-                options = GetOptions(path, options);
+            if (mode == FileMode.Open && !FileExists(path))
+            {
+                return null;
+            }
 
-                SafeFileHandle handle;
+            return new FileStream(path.Path, mode, accessMode, share, bufferSize, options);
+        }
+        
+        private Stream TryOpenFileWin(AbsolutePath path, FileAccess accessMode, FileMode mode, FileShare share, FileOptions options, int bufferSize)
+        {
+            options = GetOptions(path, options);
 
-                void openHandle()
-                {
-                    handle = NativeMethods.CreateFile(path.Path, access, share, IntPtr.Zero, mode, options, IntPtr.Zero);
-                }
+            var (result, exception) = tryOpenFile();
 
-                openHandle();
+            if (exception != null && mode == FileMode.Create &&
+                (exception.NativeErrorCode() == ERROR_ACCESS_DENIED || exception.NativeErrorCode() == ERROR_SHARING_VIOLATION))

```

Comment:
>exception.NativeErrorCode() [](start = 17, length = 27)

Is this computation cached? should we create a local to avoid double computation?

Reply:
this is a property getter and the bitmask and it is called for error case only. I was thinking of using a field here, but I think an additional local makes code harder to read.